pyscriptjs/src/exceptions.ts

```diff
@@ -0,0 +1,12 @@
+import { createAlertBanner} from "./utils"
+
+
+export class UserError extends Error {
+  constructor(message: string, showBanner =false) {
+    super(message)
+    this.name = "UserError"
+    if (showBanner) {
+      createAlertBanner(message, "error")
+    }
+  }
+}
```

Comment:
uhm, this is not what I had in mind.
What I had in mind was something like this (mostly pseudo-code but it should convey the idea):
```typescript
function withUserErrorHandler(fn) {
    try {
        fn();
    }
    catch(err) {
        if (typeof err === "UserError") // I don't know what is the right way to do typechecks 😅
            createAlertBanner(err.message);
        else
            raise;
    }
}

// main.ts
[...]
withUserErrorHandler(globalApp.main.bind(globalApp));
```

In other words:
  - UserErrors should be displayed only when they actually arrive at the top level
  - it is the caller which decides whether the user error needs to be displayed or not, not the raiser
  - it is still possible to catch `UserError` and hide it, if you need
  - if you raise `UserError`, it will NOT be displayed as a JS traceback on the console: this is good because it's not really an "internal error that we need to fix"
  - it is possible to reuse `withUserErrorHandler` in other parts of the code: this is useful e.g. if you want to execute some JS code, ensure that user errors are displayed but *do not* stop the execution if `UserError` is raised (because it will be caught be the try/catch inside `withUserErrorHandler`). Not sure how useful it will be in practice, though.

Reply:
Ah gotcha! I created the exception but didn't think about the handler, this makes sense!
Regarding the usefulness, we use a similar pattern at Coiled where a specific custom exception would mean "this is bad, show to the user, but don't blow up" and it's been useful when we want to handle things that we know may be eventually in a bad state but aren't critical :thinking:

Reply:
well, then it's a warning and not an error IMHO 😅